Implement 8-bit leopard variant

elias-orijtech opened the pull request:
MVP: straight port, with minimal differences to FF16. Passes `go test -tags noasm ./...`.

CC @klauspost @odeke-em @musalbas @liamsi

odeke-em commented on leopard8.go:
```diff
@@ -0,0 +1,1009 @@
+}
+
+func (r *leopardFF8) TotalShards() int {
+	return r.parityShards
```
This should be return r.totalShards

elias-orijtech replied:
Nice find! Fixed in this PR and in https://github.com/klauspost/reedsolomon/pull/210 for FF16.

odeke-em commented on leopard8.go:
```diff
@@ -0,0 +1,1009 @@
+
+func (r *leopardFF8) encode(shards [][]byte) error {
+	shardSize := shardSize(shards)
+	if shardSize%64 != 0 {
```
if shardSize == 0 || shardSize&64 != 0

elias-orijtech replied:
`shardSize == 0` (and other failure modes) are checked by the caller, `Encode` with `checkShards`. This is for the failure mode specific to leopard.

odeke-em commented on leopard8.go:
```diff
@@ -0,0 +1,1009 @@
+			}
+		}
+	}
+	if numberPresent == r.totalShards || !recoverAll && dataPresent == r.dataShards {
```
Minor nit: please put the second conditions in parentheses so: || (!recover ...) {

elias-orijtech replied:
Since this is a nit, I prefer to keep this identical to FF16 for now.

odeke-em reviewed (commented):
Awesome work, thank you @elias-orijtech! Just some first pass comments, please take a look.

klauspost reviewed (commented):
Let's get this to build on all platforms and we can merge and look for further improvements.

As for asm; 

`refMulAdd8` should be able to use the same asm as `galMulSliceXor`, basically just exchanging the lookup table.

`refMul8` should be `galMulSlice` with different lookup table. So plugging in asm for other platforms should be relatively trivial.

If you are up for it, you are welcome to give that a shot.

`ifftDIT48` and `ifftDIT28` will require new code, but nothing too crazy. I should be able to look at that relatively soon.

klauspost commented:
Benchmarks, with "basic" amd64 assembly:

```Go
func mulgf8(out, in []byte, log_m ffe8, o *options) {
	if o.useAVX2 {
		t := &multiply256LUT8[log_m]
		galMulAVX2_64(t[:16], t[16:32], in, out)
		done := (len(in) >> 6) << 6
		in = in[done:]
		out = out[done:]
	} else if o.useSSSE3 {
		t := &multiply256LUT8[log_m]
		galMulSSSE3(t[:16], t[16:32], in, out)
		done := (len(in) >> 4) << 4
		in = in[done:]
		out = out[done:]
	}
	out = out[:len(in)]
	mt := mul8LUTs[log_m].Value[:]
	for i := range in {
		out[i] = byte(mt[in[i]])
	}
}

// 4-way butterfly
func ifftDIT48(work [][]byte, dist int, log_m01, log_m23, log_m02 ffe8, o *options) {
	ifftDIT4Ref8(work, dist, log_m01, log_m23, log_m02, o)
}

// 4-way butterfly
func fftDIT48(work [][]byte, dist int, log_m01, log_m23, log_m02 ffe8, o *options) {
	fftDIT4Ref8(work, dist, log_m01, log_m23, log_m02, o)
}

// 2-way butterfly forward
func fftDIT28(x, y []byte, log_m ffe8, o *options) {
	// Reference version:
	mulAdd8(x, y, log_m, o)
	sliceXor(x, y, o)
}

// 2-way butterfly inverse
func ifftDIT28(x, y []byte, log_m ffe8, o *options) {
	// Reference version:
	sliceXor(x, y, o)
	mulAdd8(x, y, log_m, o)
}

func mulAdd8(out, in []byte, log_m ffe8, o *options) {
	if o.useAVX2 {
		t := &multiply256LUT8[log_m]
		galMulAVX2Xor_64(t[:16], t[16:32], in, out)
		done := (len(in) >> 6) << 6
		in = in[done:]
		out = out[done:]
	} else if o.useSSSE3 {
		t := &multiply256LUT8[log_m]
		galMulSSSE3Xor(t[:16], t[16:32], in, out)
		done := (len(in) >> 4) << 4
		in = in[done:]
		out = out[done:]
	}
	refMulAdd8(in, out, log_m)
}
```

Reconstruction:

```
BenchmarkDecode1K/4+4/cauchy-32  	  			  445640	      2548 ns/op	3215.04 MB/s	    4968 B/op	      32 allocs/op
BenchmarkDecode1K/4+4/cauchy-single-32         	  749914	      1480 ns/op	5533.66 MB/s	    1843 B/op	      27 allocs/op
BenchmarkDecode1K/4+4/leopard-gf16-32          	    1498	    782324 ns/op	  10.47 MB/s	   53274 B/op	       6 allocs/op
BenchmarkDecode1K/4+4/leopard-gf16-single-32   	    1521	    794023 ns/op	  10.32 MB/s	   50272 B/op	       5 allocs/op
BenchmarkDecode1K/4+4/leopard-gf8-32           	  271099	      4486 ns/op	1826.19 MB/s	    4120 B/op	       5 allocs/op
BenchmarkDecode1K/4+4/leopard-gf8-single-32    	   98948	     12002 ns/op	 682.54 MB/s	    1048 B/op	       2 allocs/op

BenchmarkDecode1K/8+8/cauchy-32                	  224994	      5359 ns/op	3057.37 MB/s	   10065 B/op	      53 allocs/op
BenchmarkDecode1K/8+8/cauchy-single-32         	  521190	      2242 ns/op	7307.64 MB/s	    2784 B/op	      43 allocs/op
BenchmarkDecode1K/8+8/leopard-gf16-32          	    1514	    791240 ns/op	  20.71 MB/s	   57381 B/op	      10 allocs/op
BenchmarkDecode1K/8+8/leopard-gf16-single-32   	    1504	    798729 ns/op	  20.51 MB/s	   50277 B/op	       5 allocs/op
BenchmarkDecode1K/8+8/leopard-gf8-32           	  150332	      7605 ns/op	2154.35 MB/s	    8217 B/op	       9 allocs/op
BenchmarkDecode1K/8+8/leopard-gf8-single-32    	   82618	     14574 ns/op	1124.18 MB/s	    1048 B/op	       2 allocs/op

BenchmarkDecode1K/16+16/cauchy-32              	   56107	     21262 ns/op	1541.18 MB/s	   20759 B/op	      94 allocs/op
BenchmarkDecode1K/16+16/cauchy-single-32       	  234987	      5235 ns/op	6259.80 MB/s	    5154 B/op	      75 allocs/op
BenchmarkDecode1K/16+16/leopard-gf16-32        	    1494	    796269 ns/op	  41.15 MB/s	   65562 B/op	      18 allocs/op
BenchmarkDecode1K/16+16/leopard-gf16-single-32 	    1500	    800437 ns/op	  40.94 MB/s	   50298 B/op	       6 allocs/op
BenchmarkDecode1K/16+16/leopard-gf8-32         	   77030	     14901 ns/op	2199.06 MB/s	   16411 B/op	      17 allocs/op
BenchmarkDecode1K/16+16/leopard-gf8-single-32  	   57492	     20809 ns/op	1574.70 MB/s	    1048 B/op	       2 allocs/op

BenchmarkDecode1K/32+32/cauchy-32              	    9991	    119384 ns/op	 548.95 MB/s	   44076 B/op	     175 allocs/op
BenchmarkDecode1K/32+32/cauchy-single-32       	   80649	     14371 ns/op	4560.32 MB/s	   11820 B/op	     139 allocs/op
BenchmarkDecode1K/32+32/leopard-gf16-32        	    1436	    814871 ns/op	  80.43 MB/s	   81946 B/op	      34 allocs/op
BenchmarkDecode1K/32+32/leopard-gf16-single-32 	    1482	    818185 ns/op	  80.10 MB/s	   50344 B/op	       6 allocs/op
BenchmarkDecode1K/32+32/leopard-gf8-32         	   37266	     32066 ns/op	2043.76 MB/s	   32799 B/op	      33 allocs/op
BenchmarkDecode1K/32+32/leopard-gf8-single-32  	   33559	     35680 ns/op	1836.76 MB/s	    1048 B/op	       2 allocs/op

BenchmarkDecode1K/64+64/cauchy-32              	    1468	    806659 ns/op	 162.49 MB/s	   98682 B/op	     336 allocs/op
BenchmarkDecode1K/64+64/cauchy-single-32       	   25640	     45925 ns/op	2854.05 MB/s	   32975 B/op	     267 allocs/op
BenchmarkDecode1K/64+64/leopard-gf16-32        	    1411	    875150 ns/op	 149.77 MB/s	  114812 B/op	      66 allocs/op
BenchmarkDecode1K/64+64/leopard-gf16-single-32 	    1396	    841167 ns/op	 155.82 MB/s	   50328 B/op	       7 allocs/op
BenchmarkDecode1K/64+64/leopard-gf8-32         	   16737	     71846 ns/op	1824.35 MB/s	   65603 B/op	      65 allocs/op
BenchmarkDecode1K/64+64/leopard-gf8-single-32  	   16856	     70137 ns/op	1868.81 MB/s	    1048 B/op	       2 allocs/op

BenchmarkDecode1K/128+128/cauchy-32            	     201	   5896725 ns/op	  44.46 MB/s	  237584 B/op	     657 allocs/op
BenchmarkDecode1K/128+128/cauchy-single-32     	    7476	    163281 ns/op	1605.47 MB/s	  106491 B/op	     523 allocs/op
BenchmarkDecode1K/128+128/leopard-gf16-32      	    1276	    932493 ns/op	 281.12 MB/s	  181098 B/op	     130 allocs/op
BenchmarkDecode1K/128+128/leopard-gf16-single-32         	    1342	    895151 ns/op	 292.85 MB/s	   50328 B/op	       7 allocs/op
BenchmarkDecode1K/128+128/leopard-gf8-32                 	    7730	    158086 ns/op	1658.23 MB/s	  131207 B/op	     129 allocs/op
BenchmarkDecode1K/128+128/leopard-gf8-single-32          	    7526	    147473 ns/op	1777.57 MB/s	    1048 B/op	       2 allocs/op
```

Encoding:
```
BenchmarkEncode1K/4+4/cauchy-32  	 3209847	       375.5 ns/op	21817.72 MB/s	      24 B/op	       1 allocs/op
BenchmarkEncode1K/4+4/leopard-gf16-32         	 2709714	       440.0 ns/op	18617.66 MB/s	      24 B/op	       1 allocs/op
BenchmarkEncode1K/4+4/leopard-gf8-32          	 1983574	       596.8 ns/op	13727.61 MB/s	      24 B/op	       1 allocs/op

BenchmarkEncode1K/8+8/cauchy-32               	 1000000	      1094 ns/op	14982.97 MB/s	      24 B/op	       1 allocs/op
BenchmarkEncode1K/8+8/leopard-gf16-32         	  728942	      1548 ns/op	10586.04 MB/s	      24 B/op	       1 allocs/op
BenchmarkEncode1K/8+8/leopard-gf8-32          	  782788	      1478 ns/op	11083.53 MB/s	      24 B/op	       1 allocs/op

BenchmarkEncode1K/16+16/cauchy-32             	  271365	      4355 ns/op	7524.96 MB/s	      24 B/op	       1 allocs/op
BenchmarkEncode1K/16+16/leopard-gf16-32       	  436794	      2621 ns/op	12502.37 MB/s	      24 B/op	       1 allocs/op
BenchmarkEncode1K/16+16/leopard-gf8-32        	  308798	      3747 ns/op	8744.10 MB/s	      24 B/op	       1 allocs/op

BenchmarkEncode1K/32+32/cauchy-32             	   69873	     17159 ns/op	3819.39 MB/s	      24 B/op	       1 allocs/op
BenchmarkEncode1K/32+32/leopard-gf16-32       	  136059	      8786 ns/op	7459.07 MB/s	      24 B/op	       1 allocs/op
BenchmarkEncode1K/32+32/leopard-gf8-32        	  128914	      9290 ns/op	7054.74 MB/s	      24 B/op	       1 allocs/op

BenchmarkEncode1K/64+64/cauchy-32             	   17634	     67938 ns/op	1929.28 MB/s	      39 B/op	       1 allocs/op
BenchmarkEncode1K/64+64/leopard-gf16-32       	   76540	     15587 ns/op	8408.91 MB/s	      25 B/op	       1 allocs/op
BenchmarkEncode1K/64+64/leopard-gf8-32        	   54082	     22109 ns/op	5928.32 MB/s	      26 B/op	       1 allocs/op

BenchmarkEncode1K/128+128/cauchy-32           	    4353	    268161 ns/op	 977.56 MB/s	     265 B/op	       1 allocs/op
BenchmarkEncode1K/128+128/leopard-gf16-32     	   26353	     45597 ns/op	5749.14 MB/s	      34 B/op	       1 allocs/op
BenchmarkEncode1K/128+128/leopard-gf8-32      	   23185	     51915 ns/op	5049.48 MB/s	      35 B/op	       1 allocs/op
```

elias-orijtech commented:
I've updated the PR with fixes, made it runnable without `noasm`, and added a commit with your `amd64` assembly speedups. If it's alright with you, I'll leave the `arm64` and ifft/fft optimizations to you (or future work); I'm not familiar with the asm variants, so the risk of subtle mistakes is too large.

odeke-em reviewed (approved):
LGTM, thank you @elias-orijtech, great work! Thank you @klauspost for the review and great help!
